File: apps/kotlin/androidApp/src/main/kotlin/com/runa/android/ui/components/MoonArt.kt
package com.runa.android.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Fill
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Runa moon motif, drawn with the Canvas — NO emoji, NO text glyphs. Every screen
 * that needs a moon (sign-in, onboarding, notification, diary cards/detail, and
 * the empty/offline/error states) shares these primitives so the world-view stays
 * consistent across the app and identical in spirit to the iOS MoonArt.
 */

// The moon is a FIXED cross-theme motif (design decision): it never recolors with
// the app theme, so it holds its own dark-motif constants rather than reading the
// theme tokens. Values match the original dark palette.
private val MoonLit = Color(0xFFF7F2E4)         // bright limb
private val MoonCream = Color(0xFFE8E2D0)        // sub-accent cream
private val MoonDark = Color(0xFF34343E)         // unlit disc, a touch above the surface
private val MoonRing = Color(0xFF3E3E48)
private val MoonMuted = Color(0xFF9A9AA5)        // subtle
private val MoonAccent = Color(0xFFF4A9C0)       // moonlight accent
private val MoonBackground = Color(0xFF0E0E12)   // background

/** Soft radial glow used behind the bright moons. */
private fun DrawScope.drawGlow(center: Offset, radius: Float, tint: Color, alpha: Float) {
    drawCircle(
        brush = Brush.radialGradient(
            colors = listOf(tint.copy(alpha = alpha), Color.Transparent),
            center = center,
            radius = radius,
        ),
        radius = radius,
        center = center,
    )
}

/**
 * A full, softly glowing moon — the hero mark for sign-in / onboarding / splash.
 * The whole mark (glow included) fits within [diameter].
 */
@Composable
fun GlowingMoon(
    modifier: Modifier = Modifier,
    diameter: Dp = 132.dp,
    haloTint: Color = MoonCream,
) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val moonR = size.minDimension * 0.30f
        // Outer halo, then a warmer inner glow.
        drawGlow(c, size.minDimension * 0.5f, haloTint, 0.22f)
        drawGlow(c, moonR * 1.7f, MoonLit, 0.30f)
        // Moon body with an off-centre highlight for a little dimension.
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(MoonLit, MoonCream),
                center = Offset(c.x - moonR * 0.28f, c.y - moonR * 0.28f),
                radius = moonR * 1.5f,
            ),
            radius = moonR,
            center = c,
        )
    }
}

/**
 * A small moon-phase disc for diary cards / detail. The lit fraction comes from the
 * shared moon calculator ([illumination] 0..1); [waxing] puts the lit limb on the
 * right. Rendered with the offset-shadow method: a lit disc with a same-size shadow
 * disc slid toward the dark limb, clipped to the moon.
 */
@Composable
fun MoonPhaseDisc(
    illumination: Float,
    waxing: Boolean,
    modifier: Modifier = Modifier,
    diameter: Dp = 22.dp,
) {
    Canvas(modifier.size(diameter)) {
        drawMoonPhase(illumination.coerceIn(0f, 1f), waxing)
    }
}

private fun DrawScope.drawMoonPhase(illumination: Float, waxing: Boolean) {
    val c = center
    val r = size.minDimension / 2f * 0.92f
    val moon = Path().apply { addOvalCompat(c, r) }
    // Faint halo for the bright phases so a full moon glows a little.
    if (illumination > 0.55f) drawGlow(c, r * 1.9f, MoonCream, 0.18f * illumination)
    clipPath(moon) {
        // Fully lit disc, then darken the unlit limb.
        drawCircle(MoonCream, r, c)
        val dir = if (waxing) -1f else 1f // shadow slides toward the dark side
        val shadowX = c.x + dir * 2f * r * illumination
        drawCircle(MoonDark, r, Offset(shadowX, c.y))
    }
    // Thin rim keeps the disc legible on the near-black background.
    drawCircle(MoonRing, r, c, style = Stroke(width = 1.2f))
}

/** New-moon emblem for the empty state (24): a dark disc inside a faint ring. */
@Composable
fun NewMoonEmblem(modifier: Modifier = Modifier, diameter: Dp = 108.dp) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val r = size.minDimension * 0.32f
        drawGlow(c, size.minDimension * 0.46f, MoonCream, 0.10f)
        drawCircle(MoonRing, r * 1.32f, c, style = Stroke(width = 1.4f))
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(Color(0xFF26262E), Color(0xFF14141A)),
                center = Offset(c.x - r * 0.2f, c.y - r * 0.2f),
                radius = r * 1.3f,
            ),
            radius = r,
            center = c,
        )
    }
}

/** Clouded moon for the offline state (25): a dim disc crossed by a soft stroke. */
@Composable
fun CloudedMoon(modifier: Modifier = Modifier, diameter: Dp = 108.dp) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val r = size.minDimension * 0.30f
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(Color(0xFF3A3A44), Color(0xFF24242C)),
                center = Offset(c.x - r * 0.25f, c.y - r * 0.25f),
                radius = r * 1.4f,
            ),
            radius = r,
            center = c,
        )
        drawLine(
            color = MoonMuted,
            start = Offset(c.x - r * 1.15f, c.y + r * 0.8f),
            end = Offset(c.x + r * 1.15f, c.y - r * 0.8f),
            strokeWidth = 2f,
            cap = StrokeCap.Round,
        )
    }
}

/** Gentle "stumble" emblem for the error state (27): a soft disc holding an ! mark. */
@Composable
fun StumbleEmblem(modifier: Modifier = Modifier, diameter: Dp = 108.dp) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val r = size.minDimension * 0.30f
        drawGlow(c, size.minDimension * 0.44f, MoonMuted, 0.12f)
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(Color(0xFFB9B4AC), Color(0xFF8F8B84)),
                center = Offset(c.x - r * 0.25f, c.y - r * 0.25f),
                radius = r * 1.4f,
            ),
            radius = r,
            center = c,
        )
        // ! mark, drawn (no glyph): a stem and a dot in the background colour.
        val stemTop = Offset(c.x, c.y - r * 0.42f)
        val stemBottom = Offset(c.x, c.y + r * 0.10f)
        drawLine(MoonBackground, stemTop, stemBottom, strokeWidth = r * 0.16f, cap = StrokeCap.Round)
        drawCircle(MoonBackground, r * 0.09f, Offset(c.x, c.y + r * 0.40f))
    }
}

/**
 * The notification mark (04): the glowing moon with a small moonlight-pink badge
 * carrying a minimal bell — drawn, not an emoji.
 */
@Composable
fun NotificationMoon(modifier: Modifier = Modifier, diameter: Dp = 132.dp) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val moonR = size.minDimension * 0.28f
        drawGlow(c, size.minDimension * 0.5f, MoonCream, 0.20f)
        drawGlow(c, moonR * 1.7f, MoonLit, 0.28f)
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(MoonLit, MoonCream),
                center = Offset(c.x - moonR * 0.28f, c.y - moonR * 0.28f),
                radius = moonR * 1.5f,
            ),
            radius = moonR,
            center = c,
        )
        // Pink badge up and to the right of the moon.
        val badgeC = Offset(c.x + moonR * 0.95f, c.y - moonR * 0.95f)
        val badgeR = moonR * 0.6f
        drawGlow(badgeC, badgeR * 2.1f, MoonAccent, 0.35f)
        drawCircle(MoonAccent, badgeR, badgeC)
        drawBell(badgeC, badgeR * 0.92f, MoonBackground)
    }
}

/** A minimal bell: a rounded dome body, a base line and a clapper dot. */
private fun DrawScope.drawBell(c: Offset, s: Float, color: Color) {
    val w = s * 0.9f
    val h = s * 1.0f
    val body = Path().apply {
        moveTo(c.x - w * 0.5f, c.y + h * 0.28f)
        // sides sweep up into a dome
        cubicTo(
            c.x - w * 0.5f, c.y - h * 0.15f,
            c.x - w * 0.32f, c.y - h * 0.5f,
            c.x, c.y - h * 0.5f,
        )
        cubicTo(
            c.x + w * 0.32f, c.y - h * 0.5f,
            c.x + w * 0.5f, c.y - h * 0.15f,
            c.x + w * 0.5f, c.y + h * 0.28f,
        )
        close()
    }
    drawPath(body, color, style = Fill)
    // base line + clapper
    drawLine(
        color,
        Offset(c.x - w * 0.6f, c.y + h * 0.30f),
        Offset(c.x + w * 0.6f, c.y + h * 0.30f),
        strokeWidth = s * 0.14f,
        cap = StrokeCap.Round,
    )
    drawCircle(color, s * 0.12f, Offset(c.x, c.y + h * 0.5f))
}

/**
 * The privacy-lock emblem (22): a padlock inside a soft rounded square, stroked in
 * cream — the quiet motif the lock screen and the settings screen share. Drawn, no
 * glyph.
 */
@Composable
fun LockEmblem(modifier: Modifier = Modifier, diameter: Dp = 120.dp) {
    Canvas(modifier.size(diameter)) {
        val c = center
        val boxHalf = size.minDimension * 0.30f
        val corner = boxHalf * 0.5f
        drawGlow(c, size.minDimension * 0.5f, MoonCream, 0.10f)
        // Rounded-square container.
        drawRoundRect(
            color = MoonRing,
            topLeft = Offset(c.x - boxHalf, c.y - boxHalf),
            size = Size(boxHalf * 2, boxHalf * 2),
            cornerRadius = androidx.compose.ui.geometry.CornerRadius(corner, corner),
            style = Stroke(width = 2f),
        )
        // Padlock body.
        val bodyW = boxHalf * 0.9f
        val bodyH = boxHalf * 0.72f
        val bodyTop = c.y - bodyH * 0.1f
        drawRoundRect(
            color = MoonCream,
            topLeft = Offset(c.x - bodyW / 2f, bodyTop),
            size = Size(bodyW, bodyH),
            cornerRadius = androidx.compose.ui.geometry.CornerRadius(bodyW * 0.16f, bodyW * 0.16f),
            style = Stroke(width = 2.4f),
        )
        // Shackle arc above the body.
        val shackleR = bodyW * 0.32f
        val shacklePath = Path().apply {
            moveTo(c.x - shackleR, bodyTop)
            cubicTo(
                c.x - shackleR, bodyTop - shackleR * 1.6f,
                c.x + shackleR, bodyTop - shackleR * 1.6f,
                c.x + shackleR, bodyTop,
            )
        }
        drawPath(shacklePath, MoonCream, style = Stroke(width = 2.4f))
        // Keyhole dot.
        drawCircle(MoonCream, bodyW * 0.09f, Offset(c.x, bodyTop + bodyH * 0.5f))
    }
}

private fun Path.addOvalCompat(center: Offset, radius: Float) {
    addOval(
        androidx.compose.ui.geometry.Rect(
            offset = Offset(center.x - radius, center.y - radius),
            size = Size(radius * 2, radius * 2),
        ),
    )
}

File: apps/swift/Runa/Theme/MoonArt.swift
import SwiftUI

/// Runa moon motif, drawn with SwiftUI `Canvas` — NO SF Symbols, NO emoji, NO text
/// glyphs. Mirrors the Android `MoonArt.kt` so both platforms share one world-view:
/// a glowing hero moon, a phase disc for diary records, and the empty / offline /
/// error / notification emblems.

// The moon is a FIXED cross-theme motif (design decision): it never recolors with
// the app theme, so it holds its own dark-motif constants rather than reading the
// theme tokens. Values match the original dark palette.
private let moonLit = Color(hex: 0xF7F2E4)
private let moonCream = Color(hex: 0xE8E2D0)
private let moonDark = Color(hex: 0x34343E)
private let moonRing = Color(hex: 0x3E3E48)
private let moonMuted = Color(hex: 0x9A9AA5)      // subtle
private let moonAccent = Color(hex: 0xF4A9C0)     // moonlight accent
private let moonBackground = Color(hex: 0x0E0E12) // background

private func drawGlow(_ ctx: inout GraphicsContext, center: CGPoint, radius: CGFloat, tint: Color, alpha: Double) {
    guard radius > 0 else { return }
    let rect = CGRect(x: center.x - radius, y: center.y - radius, width: radius * 2, height: radius * 2)
    let shading = GraphicsContext.Shading.radialGradient(
        Gradient(colors: [tint.opacity(alpha), .clear]),
        center: center, startRadius: 0, endRadius: radius
    )
    ctx.fill(Path(ellipseIn: rect), with: shading)
}

private func disc(_ center: CGPoint, _ r: CGFloat) -> Path {
    Path(ellipseIn: CGRect(x: center.x - r, y: center.y - r, width: r * 2, height: r * 2))
}

/// A full, softly glowing moon — the hero mark for sign-in / onboarding / splash.
struct GlowingMoon: View {
    var diameter: CGFloat = 132
    var haloTint: Color = moonCream

    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let moonR = min(size.width, size.height) * 0.30
            drawGlow(&ctx, center: c, radius: min(size.width, size.height) * 0.5, tint: haloTint, alpha: 0.22)
            drawGlow(&ctx, center: c, radius: moonR * 1.7, tint: moonLit, alpha: 0.30)
            let shading = GraphicsContext.Shading.radialGradient(
                Gradient(colors: [moonLit, moonCream]),
                center: CGPoint(x: c.x - moonR * 0.28, y: c.y - moonR * 0.28),
                startRadius: 0, endRadius: moonR * 1.5
            )
            ctx.fill(disc(c, moonR), with: shading)
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }
}

/// A small moon-phase disc for diary cards / detail. [illumination] 0..1 is the lit
/// fraction; [waxing] puts the lit limb on the right. Offset-shadow method.
struct MoonPhaseDisc: View {
    var illumination: CGFloat
    var waxing: Bool
    var diameter: CGFloat = 22

    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let r = min(size.width, size.height) / 2 * 0.92
            let f = max(0, min(1, illumination))
            if f > 0.55 { drawGlow(&ctx, center: c, radius: r * 1.9, tint: moonCream, alpha: 0.18 * Double(f)) }
            var inner = ctx
            let moon = disc(c, r)
            inner.clip(to: moon)
            inner.fill(moon, with: .color(moonCream))
            let dir: CGFloat = waxing ? -1 : 1 // shadow slides toward the dark limb
            let shadowX = c.x + dir * 2 * r * f
            inner.fill(disc(CGPoint(x: shadowX, y: c.y), r), with: .color(moonDark))
            ctx.stroke(moon, with: .color(moonRing), lineWidth: 1.2)
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }
}

/// New-moon emblem for the empty state (24): a dark disc inside a faint ring.
struct NewMoonEmblem: View {
    var diameter: CGFloat = 108
    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let r = min(size.width, size.height) * 0.32
            drawGlow(&ctx, center: c, radius: min(size.width, size.height) * 0.46, tint: moonCream, alpha: 0.10)
            ctx.stroke(disc(c, r * 1.32), with: .color(moonRing), lineWidth: 1.4)
            let shading = GraphicsContext.Shading.radialGradient(
                Gradient(colors: [Color(hex: 0x26262E), Color(hex: 0x14141A)]),
                center: CGPoint(x: c.x - r * 0.2, y: c.y - r * 0.2),
                startRadius: 0, endRadius: r * 1.3
            )
            ctx.fill(disc(c, r), with: shading)
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }
}

/// Clouded moon for the offline state (25): a dim disc crossed by a soft stroke.
struct CloudedMoon: View {
    var diameter: CGFloat = 108
    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let r = min(size.width, size.height) * 0.30
            let shading = GraphicsContext.Shading.radialGradient(
                Gradient(colors: [Color(hex: 0x3A3A44), Color(hex: 0x24242C)]),
                center: CGPoint(x: c.x - r * 0.25, y: c.y - r * 0.25),
                startRadius: 0, endRadius: r * 1.4
            )
            ctx.fill(disc(c, r), with: shading)
            var line = Path()
            line.move(to: CGPoint(x: c.x - r * 1.15, y: c.y + r * 0.8))
            line.addLine(to: CGPoint(x: c.x + r * 1.15, y: c.y - r * 0.8))
            ctx.stroke(line, with: .color(moonMuted), style: StrokeStyle(lineWidth: 2, lineCap: .round))
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }
}

/// Gentle "stumble" emblem for the error state (27): a soft disc holding an ! mark.
struct StumbleEmblem: View {
    var diameter: CGFloat = 108
    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let r = min(size.width, size.height) * 0.30
            drawGlow(&ctx, center: c, radius: min(size.width, size.height) * 0.44, tint: moonMuted, alpha: 0.12)
            let shading = GraphicsContext.Shading.radialGradient(
                Gradient(colors: [Color(hex: 0xB9B4AC), Color(hex: 0x8F8B84)]),
                center: CGPoint(x: c.x - r * 0.25, y: c.y - r * 0.25),
                startRadius: 0, endRadius: r * 1.4
            )
            ctx.fill(disc(c, r), with: shading)
            var stem = Path()
            stem.move(to: CGPoint(x: c.x, y: c.y - r * 0.42))
            stem.addLine(to: CGPoint(x: c.x, y: c.y + r * 0.10))
            ctx.stroke(stem, with: .color(moonBackground), style: StrokeStyle(lineWidth: r * 0.16, lineCap: .round))
            ctx.fill(disc(CGPoint(x: c.x, y: c.y + r * 0.40), r * 0.09), with: .color(moonBackground))
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }
}

/// The notification mark (04): the glowing moon with a small moonlight-pink badge
/// carrying a minimal drawn bell.
struct NotificationMoon: View {
    var diameter: CGFloat = 132
    var body: some View {
        Canvas { ctx, size in
            let c = CGPoint(x: size.width / 2, y: size.height / 2)
            let moonR = min(size.width, size.height) * 0.28
            drawGlow(&ctx, center: c, radius: min(size.width, size.height) * 0.5, tint: moonCream, alpha: 0.20)
            drawGlow(&ctx, center: c, radius: moonR * 1.7, tint: moonLit, alpha: 0.28)
            let shading = GraphicsContext.Shading.radialGradient(
                Gradient(colors: [moonLit, moonCream]),
                center: CGPoint(x: c.x - moonR * 0.28, y: c.y - moonR * 0.28),
                startRadius: 0, endRadius: moonR * 1.5
            )
            ctx.fill(disc(c, moonR), with: shading)

            let badgeC = CGPoint(x: c.x + moonR * 0.95, y: c.y - moonR * 0.95)
            let badgeR = moonR * 0.6
            drawGlow(&ctx, center: badgeC, radius: badgeR * 2.1, tint: moonAccent, alpha: 0.35)
            ctx.fill(disc(badgeC, badgeR), with: .color(moonAccent))
            drawBell(&ctx, center: badgeC, s: badgeR * 0.92, color: moonBackground)
        }
        .frame(width: diameter, height: diameter)
        .accessibilityHidden(true)
    }

    private func drawBell(_ ctx: inout GraphicsContext, center c: CGPoint, s: CGFloat, color: Color) {
        let w = s * 0.9, h = s * 1.0
        var body = Path()
        body.move(to: CGPoint(x: c.x - w * 0.5, y: c.y + h * 0.28))
        body.addCurve(
            to: CGPoint(x: c.x, y: c.y - h * 0.5),
            control1: CGPoint(x: c.x - w * 0.5, y: c.y - h * 0.15),
            control2: CGPoint(x: c.x - w * 0.32, y: c.y - h * 0.5)
        )
        body.addCurve(
            to: CGPoint(x: c.x + w * 0.5, y: c.y + h * 0.28),
            control1: CGPoint(x: c.x + w * 0.32, y: c.y - h * 0.5),
            control2: CGPoint(x: c.x + w * 0.5, y: c.y - h * 0.15)
        )
        body.closeSubpath()
        ctx.fill(body, with: .color(color))
        var base = Path()
        base.move(to: CGPoint(x: c.x - w * 0.6, y: c.y + h * 0.30))
        base.addLine(to: CGPoint(x: c.x + w * 0.6, y: c.y + h * 0.30))
        ctx.stroke(base, with: .color(color), style: StrokeStyle(lineWidth: s * 0.14, lineCap: .round))
        ctx.fill(disc(CGPoint(x: c.x, y: c.y + h * 0.5), s * 0.12), with: .color(color))
    }
}
